src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -513,6 +537,193 @@ internal void ReadToken(string encodedJson)
+
+                // segment between third and fourth dot (d)
+                ReadOnlySpan<char> cipherTextSpan = encodedJsonSpanCopy.Slice(0, dotIndex);
+                Dot4 = Dot3 + dotIndex + 1;
+
+                // must have something after 4th dot
+                if (dotIndex == encodedJsonSpanCopy.Length - 1)
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenMalformedException(LogMessages.IDX14310));
+
+                // segment after fourth dot (e)
+                ReadOnlySpan<char> authenticationTagSpan = encodedJsonSpanCopy.Slice(dotIndex + 1);
+                encodedJsonSpanCopy = encodedJsonSpanCopy.Slice(dotIndex + 1);
+
+                dotIndex = encodedJsonSpanCopy.IndexOf('.'); // fifth dot
+                if (dotIndex != -1)
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenMalformedException(LogMessages.IDX14122));
+                
+                if (headerSpan.IsEmpty)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogMessages.IDX14307));
+
+                if (initializationVectorSpan.IsEmpty)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogMessages.IDX14308));
+
+                if (authenticationTagSpan.IsEmpty)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogMessages.IDX14310));
+
+                if (cipherTextSpan.IsEmpty)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogMessages.IDX14306));
+
+                // right number of dots for JWE (4)
+                byte[] headerAsciiBytes = new byte[headerSpan.Length];
+                Encoding.ASCII.GetBytes(headerSpan, headerAsciiBytes);
+
+                HeaderAsciiBytes = headerAsciiBytes;
+
+                try
+                {
+                    // Discuss: Using ReadOnlySpan instead of ReadOnlyMemory
+                    // ReadOnlySpan is recommended for synchronous calls where as ReadOnlyMemory is recommended for async calls.
+                    // ReadOnlySpan performs better than ReadOnlyMemory.
+                    // https://learn.microsoft.com/en-us/dotnet/standard/memory-and-spans/memory-t-usage-guidelines
+                    Header = CreateHeaderClaimSet(Base64UrlEncoder.UnsafeDecode(headerSpan.ToArray()));  // TODO: Looking to update UnsafeDecode to accept ReadOnlySpan
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(
+                        LogMessages.IDX14102,
+                        LogHelper.MarkAsUnsafeSecurityArtifact(headerSpan.ToString(), t => t.ToString())),
+                        ex));
+                }
+
+                // dir does not have any key bytes
+                if (!secondSpan.IsEmpty)
+                {
+                    EncryptedKeyBytes = Base64UrlEncoder.UnsafeDecode(secondSpan.ToArray());
```

Comment:
will be great when we can drop the ToArray

Reply:
ToArray() is no longer in use here.